src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/base/Functions.java

```diff
@@ -22,253 +22,164 @@
-import sparksoniq.jsoniq.runtime.iterator.functions.strings.ContainsFunctionIterator;
 import sparksoniq.jsoniq.runtime.iterator.functions.strings.EndsWithFunctionIterator;
-import sparksoniq.jsoniq.runtime.iterator.functions.strings.MatchesFunctionIterator;
-import sparksoniq.jsoniq.runtime.iterator.functions.strings.NormalizeSpaceFunctionIterator;
-import sparksoniq.jsoniq.runtime.iterator.functions.strings.StartsWithFunctionIterator;
-import sparksoniq.jsoniq.runtime.iterator.functions.strings.StringJoinFunction;
-import sparksoniq.jsoniq.runtime.iterator.functions.strings.StringLengthFunctionIterator;
-import sparksoniq.jsoniq.runtime.iterator.functions.strings.SubstringAfterFunctionIterator;
-import sparksoniq.jsoniq.runtime.iterator.functions.strings.SubstringBeforeFunctionIterator;
-import sparksoniq.jsoniq.runtime.iterator.functions.strings.SubstringFunctionIterator;
-import sparksoniq.jsoniq.runtime.iterator.functions.strings.TokenizeFunctionIterator;
+import sparksoniq.jsoniq.runtime.iterator.operational.TypePromotionIterator;
 import sparksoniq.jsoniq.runtime.metadata.IteratorMetadata;
+import sparksoniq.semantics.types.ItemType;
+import sparksoniq.semantics.types.ItemTypes;
+import sparksoniq.semantics.types.SequenceType;
 import sparksoniq.spark.iterator.function.ParallelizeFunctionIterator;
-import sparksoniq.spark.iterator.function.ParseJsonFunctionIterator;
-import sparksoniq.spark.iterator.function.ParseTextFunctionIterator;
-import sparksoniq.spark.iterator.function.ParquetFileFunctionIterator;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
 
 import java.io.ByteArrayInputStream;
 import java.io.ByteArrayOutputStream;
 import java.io.ObjectInputStream;
 import java.io.ObjectOutputStream;
-import java.lang.reflect.Constructor;
-import java.util.HashMap;
-import java.util.List;
 
+import static sparksoniq.jsoniq.runtime.iterator.functions.base.Functions.FunctionNames.endsWithFunction;
+import static sparksoniq.jsoniq.runtime.iterator.functions.base.Functions.FunctionNames.hexBinaryFunction;
+import static sparksoniq.jsoniq.runtime.iterator.functions.base.Functions.FunctionNames.parallelizeFunction1;
+import static sparksoniq.jsoniq.runtime.iterator.functions.base.Functions.FunctionNames.parallelizeFunction2;
 
-import static sparksoniq.jsoniq.runtime.iterator.functions.base.Functions.FunctionNames.*;
 
 public class Functions {
-    private static HashMap<FunctionIdentifier, Class<? extends RuntimeIterator>> builtInFunctions;
+    private static final HashMap<FunctionIdentifier, BuiltinFunction> builtInFunctions;
     private static HashMap<FunctionIdentifier, FunctionItem> userDefinedFunctions;
 
+    private static final Map<String, ItemType> itemTypes;
+
+    static {
+        itemTypes = new HashMap<>();
+        itemTypes.put("String", new ItemType(ItemTypes.StringItem));
```

Comment:
Would it make sense to stick to lower-case (string, hexBinary...), to have the same syntax as in the language and avoid future confusion?

Reply:
Yes, thanks for spotting this.